src/asmcnc/keyboard/custom_keyboard.py

```diff
@@ -110,14 +113,44 @@ def key_up(self, keycode, internal, modifiers):
                 if keycode == "Han/Yeong":
                     #https://en.wikipedia.org/wiki/Language_input_keys#Keys_for_Korean_Keyboards
                     self.layout = self.qwertyKR_layout if self.layout == self.kr_layout else self.kr_layout
+                    self.previous_layout = self.layout
                 if keycode == "escape":
                     self.text_instance.focus = False
                 if keycode == "backspace":
                     self.text_instance.do_backspace()
-                return
 
+                if keycode == "layout":
+                    self.layout = self.numeric_layout if self.layout == self.previous_layout else self.previous_layout
+
+                    if self.layout == self.numeric_layout:
+                        self.width = Window.width/3
+                    else:
+                        self.width = Window.width
+
+                    self.set_keyboard_background()
+
+                    # Make sure keyboard never goes off-screen and becomes unusable/unreachable
+                    if self.pos[0] + self.width > Window.width:
+                        self.pos = (Window.width-self.width, self.pos[1])
+                    if self.pos[1] < 0:
+                        self.pos = (self.pos[0], 0)
+                    if self.pos[0] < 0:
+                        self.pos = (0, self.pos[1])
+                    if self.pos[1] + self.height > Window.height:
+                        self.pos = (self.pos[0], Window.height - self.height)
+                return
             self.text_instance.insert_text(internal)
 
+    def set_keyboard_background(self):
+        if self.do_translation == (True,True):
```

Comment:
why is this a tuple, when further up it's just set to 'True'?

Reply:
- I will change the layout of the numpad, I didn't realise it's quite stupid at the moment.
- Since the user will never get the numpad on its own for the moment, I don't think so. I've also talked to Ed about it and he said he doesn't think there's any instance where there isn't an "enter" button on the screen for any text inputs. Adding one is really easy though, functionality is the same as normal enter. Would you like me to add one?
- I don't know why kivy treats it as a tuple either and it took a little while to realise that it's not the same as when it's set. I assume `do_translation` is stored as a tuple because it has a boolean for vertical and horizontal translation.

Reply:
Is this okay? 

![image](https://github.com/YetiTool/easycut-smartbench/assets/126763092/bbfa9ba9-1d00-4bc7-8138-82a6cd289ebb)